test/e2e/pkg/sourceclient_test.go

```diff
@@ -0,0 +1,208 @@
+package e2e_test
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"time"
+
+	jsonpatch "github.com/evanphx/json-patch"
+	. "github.com/onsi/ginkgo/v2"
+	. "github.com/onsi/gomega"
+
+	"github.com/openshift-online/maestro/pkg/client/cloudevents/grpcsource"
+
+	"k8s.io/apimachinery/pkg/api/meta"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/util/rand"
+	"k8s.io/apimachinery/pkg/watch"
+
+	workv1 "open-cluster-management.io/api/work/v1"
+
+	"open-cluster-management.io/sdk-go/pkg/cloudevents/generic/options/grpc"
+	"open-cluster-management.io/sdk-go/pkg/cloudevents/work"
+	"open-cluster-management.io/sdk-go/pkg/cloudevents/work/common"
+	"open-cluster-management.io/sdk-go/pkg/cloudevents/work/source/codec"
+)
+
+var _ = Describe("gRPC Source ManifestWork Client Test", func() {
+	Context("Watch work status with gRPC source ManifestWork client", func() {
+		var ctx context.Context
+		var cancel context.CancelFunc
+
+		var sourceID string
+		var grpcOptions *grpc.GRPCOptions
+
+		var watchedWorks []*workv1.ManifestWork
+
+		BeforeEach(func() {
+			ctx, cancel = context.WithCancel(context.Background())
+
+			sourceID = "sourceclient-test" + rand.String(5)
+
+			watchedWorks = []*workv1.ManifestWork{}
+
+			grpcOptions = grpc.NewGRPCOptions()
+			grpcOptions.URL = grpcServerAddress
+
+			workClient, err := work.NewClientHolderBuilder(grpcOptions).
+				WithClientID(fmt.Sprintf("%s-watcher", sourceID)).
+				WithSourceID(sourceID).
+				WithCodecs(codec.NewManifestBundleCodec()).
+				WithWorkClientWatcherStore(grpcsource.NewRESTFullAPIWatcherStore(apiClient, sourceID)).
+				WithResyncEnabled(false).
+				NewSourceClientHolder(ctx)
+			Expect(err).ShouldNot(HaveOccurred())
+
+			watcher, err := workClient.ManifestWorks(consumer_name).Watch(ctx, metav1.ListOptions{})
+			Expect(err).ShouldNot(HaveOccurred())
+
+			go func() {
+				ch := watcher.ResultChan()
+				for {
+					select {
+					case <-ctx.Done():
+						return
+					case event, ok := <-ch:
+						if !ok {
+							return
+						}
+						switch event.Type {
+						case watch.Modified:
+							if work, ok := event.Object.(*workv1.ManifestWork); ok {
+								watchedWorks = append(watchedWorks, work)
+							}
+						case watch.Deleted:
+							if work, ok := event.Object.(*workv1.ManifestWork); ok {
+								watchedWorks = append(watchedWorks, work)
+							}
+						}
+					}
+				}
+			}()
+		})
+
+		AfterEach(func() {
+			cancel()
+		})
+
+		It("The work status should be watched", func() {
+			workClient, err := work.NewClientHolderBuilder(grpcOptions).
+				WithClientID(fmt.Sprintf("%s-client", sourceID)).
+				WithSourceID(sourceID).
+				WithCodecs(codec.NewManifestBundleCodec()).
+				WithWorkClientWatcherStore(grpcsource.NewRESTFullAPIWatcherStore(apiClient, sourceID)).
+				WithResyncEnabled(false).
```

Comment:
is it necessary to initiate the watcher store for no-watcher work client?

Reply:
yes, it's is required, in addition to implementing the watcher interface, this store also implements get/list for work client